File: Ch. 07/causal_forests/code/ch7_causal_forest_u_program.py
# =============================================================================
# Chapter 7: Causal Forest Analysis for U-Program (GRF implementation)
# =============================================================================
# Estimates Conditional Average Treatment Effects (CATEs) using causal forests.
# Analyzes the effect of Math curriculum allocation on disciplinary infractions
# during the 16-17 academic year across 4 quarters.
#
# Uses EconML's GRF CausalForest to estimate heterogeneous treatment effects
# based on pre-treatment characteristics including ScanQuest scores,
# demographics, and school variables.
#
# econml.grf.CausalForest is a Python implementation of Generalized Random Forests,
# providing the same algorithm as R's grf package (Athey, Tibshirani, Wager 2019).
# This solves the same local moment equation problem as R's causal_forest.
#
# Reference: Chapter 7, Heterogeneous Treatment Effects

from pathlib import Path
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from econml.grf import CausalForest
import statsmodels.api as sm


# --- Setup -------------------------------------------------------------------
# Automatically resolve paths relative to this script's location.
SCRIPT_DIR = Path(__file__).resolve().parent
DATA_DIR = SCRIPT_DIR.parent / "data"
OUTPUT_DIR = SCRIPT_DIR.parent / "output"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)


# --- Load and filter data ----------------------------------------------------
data_up = pd.read_csv(DATA_DIR / "u_program_data.csv")

print("\n" + "=" * 80)
print("CHAPTER 7: Causal Forest Analysis for U-Program (GRF)")
print("=" * 80)
print("The program took place in the 16-17 academic year, spread in 4 different quarters.")
print("=" * 80 + "\n")

# Filter to Math treatment and control groups
data_up_math_control = data_up[data_up['treatment'].isin(['Math', 'control'])].copy()


# --- Select pre-treatment characteristics ------------------------------------
# Select all pre-treatment variables for the causal forest
pre_treatment_cols = [col for col in data_up_math_control.columns if col.startswith('ScanQuest') and col.endswith('PRE')]
pre_treatment_cols += ['school', 'grade', 'class', 'age', 'female', 'white', 'black',
                       'hispanic', 'other_race', 'disc_1415', 'disc_1516',
                       'DaysAbsent_1415', 'DaysAbsent_1516', 'enrollment_1415',
                       'enrollment_1516']

pre_treatment_characs = data_up_math_control[pre_treatment_cols].copy()


# --- Check for missing data --------------------------------------------------
na_proportions = pre_treatment_characs.isna().mean()
many_missing = na_proportions[na_proportions > 0.05]

print("Variables with >5% missing data:")
if len(many_missing) > 0:
    print(many_missing)
else:
    print("None")
print()

# Remove variables with excessive missing data
pre_treatment_characs = pre_treatment_characs.drop(
    columns=['disc_1415', 'DaysAbsent_1415'], errors='ignore'
)


# --- Create post-treatment outcome variable ----------------------------------
data_up_math_control['disc_post'] = data_up_math_control.apply(
    lambda row: row[f'disc_q{int(row["treated_quarter"])}_1617']
    if pd.notna(row['treated_quarter']) else np.nan,
    axis=1
)


# --- Estimate baseline treatment effect --------------------------------------
# Create treatment indicator (1 = Math, 0 = control)
data_up_math_control['treatment_Math'] = (
    data_up_math_control['treatment'] == 'Math'
).astype(int)

# Baseline OLS regression
m2b = sm.OLS.from_formula(
    'disc_post ~ treatment_Math',
    data=data_up_math_control
).fit()

print("Baseline treatment effect (OLS regression):")
print(m2b.summary())
print(f"\nThe children that were allocated to the Math have {-m2b.params['treatment_Math']:.4f} "
      f"less disciplinary infractions.\n")


# --- Prepare data for causal forest ------------------------------------------
# Remove rows with missing values in outcome, treatment, or covariates
# This ensures R and Python use the same data
complete_cases = (
    data_up_math_control['disc_post'].notna() &
    data_up_math_control['treatment_Math'].notna() &
    pre_treatment_characs.notna().all(axis=1)
)

print(f"Complete cases: {complete_cases.sum()} out of {len(complete_cases)} observations\n")

X = pre_treatment_characs[complete_cases].values
Y = data_up_math_control.loc[complete_cases, 'disc_post'].values
W = data_up_math_control.loc[complete_cases, 'treatment_Math'].values


# --- Estimate causal forest --------------------------------------------------
print("Fitting causal forest using econml.grf.CausalForest (GRF algorithm)...")

# Initialize CausalForest
# This implements the Generalized Random Forest algorithm, matching R's grf package
# It solves: E[(Y - <theta(x), T> - beta(x)) (T;1) | X=x] = 0
causal_forest = CausalForest(
    n_estimators=2000,              # Same as R's num.trees default
    criterion='mse',                # Mean squared error criterion (default)
    min_samples_leaf=5,             # Same as R's min.node.size default
    max_depth=None,                 # Unlimited depth (same as R)
    max_samples=0.5,                # Sample fraction (R's default is 0.5)
    min_balancedness_tol=0.45,      # Enforces balanced splits (R default)
    honest=True,                    # Use honest splitting (GRF default)
    inference=True,                 # Enable confidence intervals
    fit_intercept=True,             # Fit intercept beta(x) as nuisance parameter
    random_state=42,                # For reproducibility
    n_jobs=-1                       # Use all CPU cores
)

# Fit the causal forest
# GRF uses honest splitting: data split for building trees vs. estimating effects
causal_forest.fit(X, T=W, y=Y)

print("Causal forest fitted successfully.")


# --- Extract predictions (CATEs) ---------------------------------------------
# Predict individual treatment effects (CATEs)
# Note: predict() returns shape (n_samples, n_treatments), so we flatten for single treatment
tau_hat = causal_forest.predict(X).flatten()

df_cdf = pd.DataFrame({'tau_hat': tau_hat})


# --- Create CATE cumulative relative frequency plot -------------------------
plt.figure(figsize=(10, 6))

# Sort CATE values and calculate cumulative relative frequencies
tau_sorted = np.sort(tau_hat)
cumulative_freq = np.arange(1, len(tau_sorted) + 1) / len(tau_sorted)

# Plot cumulative relative frequency
plt.plot(tau_sorted, cumulative_freq, linewidth=1.2, color='#1f77b4')
plt.grid(True, alpha=0.3, linestyle='--', linewidth=0.5)

plt.xlabel('CATE', fontsize=11)
plt.ylabel('Cumulative Relative Frequency', fontsize=11)
plt.title('CATEs (econml.grf.CausalForest - GRF Algorithm)', fontsize=13, fontweight='bold')
plt.figtext(0.5, 0.01,
            'Effect of being allocated to the Math Curriculum on number of disciplinary infractions.',
            ha='center', fontsize=9, style='italic')

# Save plot
output_file = OUTPUT_DIR / 'ch7_causal_forest_cate_cumulative_py.png'
plt.savefig(output_file, dpi=300, bbox_inches='tight')
plt.close()


# --- Print results -----------------------------------------------------------
print("=" * 80)
print("CATE Analysis Results (econml.grf.CausalForest)")
print("=" * 80)
print(f"Saved to: {output_file}")
print(f"Number of observations: {len(tau_hat)}")
print(f"Mean CATE: {tau_hat.mean():.4f}")
print(f"Median CATE: {np.median(tau_hat):.4f}")
print(f"Min CATE: {tau_hat.min():.4f}")
print(f"Max CATE: {tau_hat.max():.4f}")
print("\nAll CATE values are negative, which means that no one")
print("is predicted to increase disciplinary infractions due to treatment.")
print("=" * 80)

# =============================================================================
# END OF CHAPTER 7: Causal Forest Analysis (econml.grf.CausalForest)
# =============================================================================

File: Ch. 07/causal_forests/code/ch7_causal_forest_u_program.R
# =============================================================================
# Chapter 7: Causal Forest Analysis for U-Program
# =============================================================================
# Estimates Conditional Average Treatment Effects (CATEs) using causal forests.
# Analyzes the effect of Math curriculum allocation on disciplinary infractions
# during the 16-17 academic year across 4 quarters.
#
# Uses generalized random forests (grf) to estimate heterogeneous treatment
# effects based on pre-treatment characteristics including ScanQuest scores,
# demographics, and school variables.
#
# Reference: Chapter 7, Heterogeneous Treatment Effects

# --- Setup -------------------------------------------------------------------
rm(list = ls())

# Load required libraries
library(tidyverse)
library(readr)
library(grf)

# Automatically set working directory to the folder containing this script.
# Supports RStudio (Source/Knit), source(), and Rscript from the command line.
if (requireNamespace("rstudioapi", quietly = TRUE) && rstudioapi::isAvailable()) {
  setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
} else if (length(sys.frames()) > 0 && !is.null(sys.frame(1)$ofile)) {
  setwd(dirname(sys.frame(1)$ofile))
} else if (!is.null(commandArgs(trailingOnly = FALSE))) {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("^--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    setwd(dirname(normalizePath(sub("^--file=", "", file_arg))))
  }
}

# Define paths relative to script location
data_dir <- file.path(dirname(getwd()), "data")
output_dir <- file.path(dirname(getwd()), "output")
dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)


# --- Load and filter data ----------------------------------------------------
data_up <- read_csv(file.path(data_dir, "u_program_data.csv"))

cat("\n", strrep("=", 80), "\n", sep = "")
cat("CHAPTER 7: Causal Forest Analysis for U-Program\n")
cat(strrep("=", 80), "\n", sep = "")
cat("The program took place in the 16-17 academic year, spread in 4 different quarters.\n")
cat(strrep("=", 80), "\n\n", sep = "")

# Filter to Math treatment and control groups
data_up_math_control <- data_up %>% filter(treatment %in% c("Math", "control"))


# --- Select pre-treatment characteristics ------------------------------------
pre_treatment_characs <-
  data_up_math_control %>%
  dplyr::select(matches("^ScanQuest.*PRE$"), school, grade, class, age, female, white, black, hispanic, other_race, disc_1415, disc_1516, DaysAbsent_1415,
         DaysAbsent_1516, enrollment_1415, enrollment_1516)


# --- Check for missing data --------------------------------------------------
na_table <-
  pre_treatment_characs %>%
  dplyr::summarise(across(everything(), ~ mean(is.na(.)))) %>%
  tidyr::pivot_longer(
    cols = everything(),
    names_to = "variable",
    values_to = "na_proportion"
  )

many_missing <- na_table %>% filter(na_proportion > 0.05)

cat("Variables with >5% missing data:\n")
print(many_missing)
cat("\n")

# Remove variables with excessive missing data
pre_treatment_characs <-
  pre_treatment_characs %>%
  dplyr::select(-disc_1415, -DaysAbsent_1415)


# --- Create post-treatment outcome variable ----------------------------------
data_up_math_control <- data_up_math_control %>%
  mutate(disc_post = case_when(treated_quarter == 1 ~ disc_q1_1617,
                               treated_quarter == 2 ~ disc_q2_1617,
                               treated_quarter == 3 ~ disc_q3_1617,
                               treated_quarter == 4 ~ disc_q4_1617))


# --- Estimate baseline treatment effect --------------------------------------
# Create treatment indicator (1 = Math, 0 = control)
data_up_math_control <- data_up_math_control %>%
  mutate(treatment_Math = (treatment == "Math"))

m2b <- data_up_math_control %>% lm(disc_post ~ treatment_Math, data = .)

cat("Baseline treatment effect (OLS regression):\n")
summary(m2b)
cat(sprintf("\nThe children that were allocated to the Math have %.4f less disciplinary infractions.\n\n",
            -round(m2b$coefficients[2], 4)))


# --- Prepare data for causal forest ------------------------------------------
# Remove rows with missing values in outcome, treatment, or covariates
# This ensures R and Python use the same data
complete_cases <- complete.cases(pre_treatment_characs) &
                  !is.na(data_up_math_control$disc_post) &
                  !is.na(data_up_math_control$treatment_Math)

cat(sprintf("Complete cases: %d out of %d observations\n\n",
            sum(complete_cases), length(complete_cases)))

X <- pre_treatment_characs[complete_cases, ]
Y <- data_up_math_control$disc_post[complete_cases]
W <- data_up_math_control$treatment_Math[complete_cases]


# --- Estimate causal forest --------------------------------------------------
causal_forest_disc_math <- causal_forest(X, Y, W)


# --- Extract predictions -----------------------------------------------------
tau_hat <- predict(causal_forest_disc_math)$predictions

causal_forest_disc_math$predictions

df_cdf <- data.frame(tau_hat = tau_hat)


# --- Create CATE cumulative relative frequency plot -------------------------
# Sort CATE values and calculate cumulative relative frequencies
df_cdf_sorted <- df_cdf %>%
  arrange(tau_hat) %>%
  mutate(cumulative_freq = row_number() / n())

cate_plot <- ggplot(df_cdf_sorted, aes(x = tau_hat, y = cumulative_freq)) +
  geom_line(linewidth = 1.2, color = "#1f77b4") +
  labs(
    x = "CATE",
    y = "Cumulative Relative Frequency",
    title = "CATEs",
    caption = "Effect of being allocated to the Math Curriculum on number of disciplinary infractions."
  ) +
  theme_minimal(base_size = 14) +
  theme(
    panel.grid.minor = element_line(linetype = "dashed", linewidth = 0.3, color = "gray80"),
    panel.grid.major = element_line(linetype = "dashed", linewidth = 0.3, color = "gray70"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )


# --- Save output -------------------------------------------------------------
output_file <- file.path(output_dir, "ch7_causal_forest_cate_cumulative_r.png")
ggsave(output_file, plot = cate_plot, width = 10, height = 6, units = "in", dpi = 300)

cat(strrep("=", 80), "\n", sep = "")
cat("CATE Analysis Results\n")
cat(strrep("=", 80), "\n", sep = "")
cat("Saved to:", output_file, "\n")
cat(sprintf("Number of observations: %d\n", length(tau_hat)))
cat(sprintf("Mean CATE: %.4f\n", mean(tau_hat)))
cat(sprintf("Median CATE: %.4f\n", median(tau_hat)))
cat(sprintf("Min CATE: %.4f\n", min(tau_hat)))
cat(sprintf("Max CATE: %.4f\n", max(tau_hat)))
cat("\nAll CATE values are negative, which means that no one\n")
cat("is predicted to increase disciplinary infractions due to treatment.\n")
cat(strrep("=", 80), "\n", sep = "")

# =============================================================================
# END OF CHAPTER 7: Causal Forest Analysis
# =============================================================================
